 ## 1  Описание проекта

   Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Необходимо проанализировать базу данных, которая содержит информацию о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Цель исследования: 

Проанализировать базу данных для формирования ценностного предложения для нового продукта

## 2 Описание данных 

  Таблица books содержит данные о книгах:
  
* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя  

 Таблица authors содержит данные об авторах:

* author_id — идентификатор автора;
* author — имя автора

 Таблица publishers содержит данные об издательствах:
 
* publisher_id — идентификатор издательства;
* publisher — название издательства;

 Таблица ratings:
 
* Содержит данные о пользовательских оценках книг:
* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

 Таблица reviews: 
* содержит данные о пользовательских обзорах на книги:
* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя пользователя, написавшего обзор;
* text — текст обзора. 



## 3 Загрузка данных  и изучение общей информации

In [1]:

# импортируем библиотеки
import pandas as pd
import sqlalchemy as sa

# устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})
# чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    '''Открываем соединение, получаем данные из sql, закрываем соединение'''
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)
                           


Изучим таблицы базы данных

In [2]:
# посмотрим на первые пять строк таблицы books
query = '''SELECT * FROM books LIMIT 5'''
get_sql_data(query)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
# посчитаем количество строк в таблице books
query = ''' SELECT * FROM books'''
get_sql_data(query)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

Всего в таблице books 6 столбцов и 1000 строк.

In [4]:
# посчитаем количество строк в таблице authors
query = ''' SELECT * FROM authors'''
get_sql_data(query)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

В таблице authors 636 строк и 2 столбца 

In [5]:
# выведем первые пять строк в таблице authors
query = ''' SELECT * FROM authors LIMIT 5'''
get_sql_data(query)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
# посчитаем количество строк в таблице ratings
query = ''' SELECT * FROM ratings'''
get_sql_data(query)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

Всего в таблице ratings 6456 строк и 4 столбца 

In [7]:
# выведем первые пять строк таблицы ratings 
query = ''' SELECT * FROM ratings LIMIT 5'''
get_sql_data(query)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [8]:
# посчитаем количество строк в таблице reviews 
query = ''' SELECT * FROM reviews'''
get_sql_data(query) 


review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

Всего в таблице reviews 2793 строки и 4 столбца 

In [9]:
# выведем первые пять строк в таблице reviews
query = ''' SELECT * FROM reviews LIMIT 5'''
get_sql_data(query) 



review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [10]:
# посчитаем количество строк в таблице publishers
query = ''' SELECT * FROM publishers'''
get_sql_data(query)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

Всего в таблице 340 строк и 2 столбца 

In [11]:
# выведем первые пять строк таблицы publishers
query = ''' SELECT * FROM publishers LIMIT 5'''
get_sql_data(query) 


publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

# 4  Запросы к базе данных 

1. Посчитайте, сколько книг вышло после 1 января 2000 года 

In [12]:
# формируем запрос 
query = ''' SELECT COUNT(*) FROM books
            WHERE publication_date::date > '2000-01-01' '''
            
# выводим результат 
get_sql_data(query) 

count
0    819

 После 1 января 2020 года вышло 819 книг 
    

2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [13]:
# формируем запрос  
query = ''' SELECT title,
                   COUNT (DISTINCT review_id) AS review_count,
                   ROUND(AVG(rating), 2) AS avg_rating
            FROM books AS b 
            LEFT JOIN ratings AS r ON b.book_id = r.book_id 
            LEFT JOIN reviews AS rev ON b.book_id = rev.book_id
            GROUP BY b.book_id
            ORDER BY avg_rating DESC, review_count DESC '''
            
        

# выводим результат 
get_sql_data(query)




title  review_count  \
0                        A Dirty Job (Grim Reaper  #1)             4   
1              School's Out—Forever (Maximum Ride  #2)             3   
2         Moneyball: The Art of Winning an Unfair Game             3   
3                   The Big Bad Wolf  (Alex Cross  #9)             2   
4                 Welcome to Temptation (Dempseys  #1)             2   
..                                                 ...           ...   
995  The World Is Flat: A Brief History of the Twen...             3   
996                                      Drowning Ruth             3   
997                  His Excellency: George Washington             2   
998                                              Junky             2   
999                               Harvesting the Heart             2   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

   В нашем распоряжении книги, отсортирвоанные по количеству обзоров и по среднему рейтингу. Книга с наибольшим средним рейтингом и количеством обзоров - это A Dirty Job (Grim Reaper #1) 

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [14]:
# формируем запрос
query = ''' WITH publisher_count_books AS
            (SELECT publisher, 
                    COUNT(book_id) AS book_count 
             FROM publishers AS p 
             INNER JOIN books AS b ON p.publisher_id = b.publisher_id 
             WHERE num_pages > 50 
             GROUP BY publisher
             ORDER BY book_count DESC) 

             SELECT publisher
             FROM publisher_count_books
             WHERE book_count =  (SELECT MAX(book_count) 
                                 FROM publisher_count_books) '''
                                          
             
             
                      
# выводим результат                                     
get_sql_data(query) 


publisher
0  Penguin Books

 Таким образом, 'Penguin Books' является издательством, которое выпустило наибольшее число книг толще 50 страниц

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [15]:
# формируем запрос
query = ''' SELECT author, 
                   round(AVG(rating),2) AS avg_rating
            FROM authors AS a 
            JOIN books AS b ON a.author_id = b.author_id
            JOIN ratings AS r ON b.book_id = r.book_id
            WHERE b.book_id IN ( SELECT book_id 
                                FROM ratings 
                                GROUP BY book_id
                                HAVING COUNT(rating_id) >= 50 
                                ORDER BY COUNT(rating_id) DESC) 
            GROUP BY author 
            ORDER BY avg_rating DESC
            LIMIT 1 '''                    
            




# выводим результат 
get_sql_data(query)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Автором с самой выcокой средней оценкой книг (4.29) среди книг с 50 и более оценками - является J.K.rowling / Mary GrandPre

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [16]:
# формируем запрос
query = '''  SELECT ROUND(AVG(review_cnt)) 
             FROM ( SELECT username, 
                           COUNT(review_id) AS review_cnt 
                    FROM reviews 
                    WHERE username IN  ( SELECT username 
                                         FROM ratings 
                                         GROUP BY username 
                                         HAVING COUNT(rating_id) > 48 ) 
             GROUP BY username 
             ) AS review_cnt  '''
                                     
            
            

# выводим результат 
get_sql_data(query)

round
0   24.0

Таким образом, среднее количество обзоров от пользователей, которые поставили больше 48 оценок, равно 24